### ISSUE TITLE
Implementing createURL is too hard

### ISSUE THREAD

**Issue by briansmith** [WICG/trusted-types#151]
Nowhere is an example of `createURL` given: not in the draft spec, nor the blog post, nor in the demo.

I think that the lack of `createURL` consideration makes some of the current examples misleading. For example, in https://wicg.github.io/trusted-types/dist/spec/#policies we have a policy that's described as "Library initialized with a policy allowing it to load additional scripts from a given host". However, is the intent of the policy to also block the dynamic creation of link `<a href=https://example.com>link</a>`? Probably not.

It would be especially useful to see an example of a policy that allows the creation of a `TrustedURL` for any URL that can't result in script execution, without allowing the creation of any orther object of a trusted type. In other words, I am requesting an example that emulates `Content-Security-Policy: script-src 'none'`.  I think a lot of people who don't use templating frameworks based on string interpolation would use such a policy.

**Comment by briansmith** [comment]
BTW, I titled this issue "Implementing `createURL` is too hard" because my hypothesis is that creating the kind of example would be error-prone and/or would require a lot of tedious coding.

**Comment by koto** [comment]
For controlling script loading, you need a policy for `TrustedScriptURL` only, not `TrustedURL`. So a policy that's equivalent to `script-src 'none'` (in DOM only - Trusted Types don't apply for something that's already in the document e.g. inserted by HTML parser or reflected from a server) would be:

```js
TrustedTypes.createPolicy('foo', {
  createScriptURL(_) {
    throw 1;
  }
}) 
```

or a policy that does not define `createScriptURL` function at all. It's not an exact equivalent (e.g. javascript URLs are allowed for links), as this property would have to hold in all the policies created for a document. 

The `cdn-scripts` policy in the [spec example](https://wicg.github.io/trusted-types/dist/spec/#policies) would block `a.href` link URL creation because it doesn't have a `createURL` function - it's a policy that gives a limited capability to only create script URLs. You can have multiple policies in the document, as most likely you want your different application parts to have different capabilities. 

As for `createURL` is hard, so far we've been using:

```js
  createURL(i) {
    const url = new URL(i, document.baseURI);
    if (['http:', 'https:'].includes(url.protocol)) {
      return i;
    }
    throw new TypeError('Disallowed URL');
  },
```

which is most likely what most authors want (i.e. link to anything, but disable `javascript:` URLs).

As a general note, we don't want to make the spec itself heavy with examples and recommendations on building policies, but are focusing now on creating JS libraries that create useful policies & documentation on how to integrate TT with web applications.

**Comment by briansmith** [comment]
> So a policy that's equivalent to `script-src 'none'` ... would be:

That's not equivalent to `script-src 'none'` because it blocks many non-script things. Sorry I wasn't clear: I mean, I want to see the TrustedTypes police that's equivalent to a CSP that consists *only* of `script-src 'none'`. In particular, it is unclear (based on my experience at least) that, without additional work, it is impossible to create a link `<a href>`.

> if (['http:', 'https:'].includes(url.protocol)) {

This seems reasonable for `<a href>` but it's less reasonable for subresources like `<img src>`; one would not want to support `http` for subresources, generally. I think it's fair to say that these distinctions are out of scope of Trusted Types as the given scope is prevention of DOM XSS.

> which is most likely what most authors want (i.e. link to anything, but disable javascript: URLs).

The given policy blocks `data:`, `blob:`, `about:`, and `ftp:` at least.

Let's forget about `about:` and `ftp:` for now. In order to allow `data:` and `blob:`, one has to do a lot more  work because there aren't separate types for image URLs. Consider:

```
<img style='width: 100px; height: 100px' src="data:image/svg+xml,<svg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 512 512'><path d='M224 20387.814V512L32 320l192-192v126.912C447.375 387.814z'/><script>alert(1)</script></svg>">
```

If we had a separate `createImageURL: DOMString -> TrustedImageURL` then we could use a check similar to yours:

> if (['data:', 'http:', 'https:'].includes(url.protocol)) {

This would work fine (AFAICT) because scripts within images embedded using `<img src>` will not run.

However, without such a separate type, we have to do a lot more work, because the URL may be used for `<a href>` or `<iframe src>`:

```
<a href="data:image/svg+xml,<svg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 512 512'><path d='M224 20387.814V512L32 320l192-192v126.912C447.375 387.814z'/><script>alert(1)</script></svg>">link</a>
```

Thus, the TrustedTypes policy would need to sanitize `data:` URLs similar to the way it would implement in `createHTML` in order to allow embedding SVG via `<img src>`. That is a lot more work, especially if the framework doesn't otherwise need to implement `createHTML`.

> As a general note, we don't want to make the spec itself heavy with examples and recommendations on building policies, but are focusing now on creating JS libraries that create useful policies & documentation on how to integrate TT with web applications.

Understood. Also, it is a fair point that for many applications, whitelisting HTTP and HTTPS based on the scheme is probably good enough. I think it's unfortunate that going from that to also allowing `<img src='data:image/svg+xml,...>` seems to require a disproportionate leap of effort. Like I mentioned above, if navigation (`<a href>`), image (`<img src>`), and other subresource URLs had separate types, less effort would be required, so maybe reconsidering using one type for every kind of non-script-document URL is worth exploring.

**Comment by koto** [comment]
I don't think we want to go to the extreme of CSP, trying to guard each of the sinks separately (e.g. URLs for fonts, images, links, XHR etc) as the [focus](https://wicg.github.io/trusted-types/dist/spec/#goals) is to prevent injections that lead to DOM XSS (i.e. not being an equivalent of a programmatic CSP). Perhaps we should make this more clear in the non-goals section.

In modern browsers, as far as DOM XSS goes, you should be OK to allow `data:` protocol for `TrustedURL`s. As far as I can tell, `data:` URLs execute in empty origin, so the JS execution through a `data:` iframe just as if you've loaded an external document, that uses its own scripts.

```
data:text/html,parentdoc<iframe src="data:image/svg+xml,<svg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 512 512'><path d='M224 20387.814V512L32 320l192-192v126.912C447.375 387.814z'/><script>alert(top.location.href)</script></svg>"></iframe>
```

Based on that, is there an increased DOM XSS risk by treating `iframe.src` sink the same as, say, `img.src`? Subresources are indeed different than navigation, but we want to keep the number of types minimal for simplicity. And, as such, [`TrustedScriptURL`](https://wicg.github.io/trusted-types/dist/spec/#trused-script-url) points to script resources like JS code or plugins, and [`TrustedURL`](https://wicg.github.io/trusted-types/dist/spec/#trusted-url) points to everything else. We're most likely add a type for stylesheet URLs to the list #104, but don't really want to extend the list without a convincing argument. 

On your other points, you can use multiple policies in a document. Examples only show policies built for a single purpose (e.g. restricting script URLs). Those policy objects could not be used to create URLs for `a.href`, but could be extended to have that capability, but perhaps a better idea is to use multiple policy objects, delivered to different parts of your application. For example, your templating library should probably not be able to create script URLs, but your script loaders should not create HTML.

**Comment by mikesamuel** [comment]
> I don't think we want to go to the extreme of CSP, trying to guard each of the sinks separately (e.g. URLs for fonts, images, links, XHR etc) as the focus is to prevent injections that lead to DOM XSS (i.e. not being an equivalent of a programmatic CSP). Perhaps we should make this more clear in the non-goals section.

Agreed, but if we were to add one more level of granularity to URLs, I'd vote for an equivalent of CSP form-action since it allows exfil of nonces and passwords without having to do extra work to cause them to be carried along.

**Comment by briansmith** [comment]
I did try to think through whether it makes sense to have more types, focusing on `<img src>` for SVG. My conclusion was that it doesn't really make sense to treat the restricted handling of `<img src>` for SVG as a security feature so it doesn't make sense to give it a separate type in this framework.

One thing that threw me off a bit was the claim in the documentation that all the security-critical logic is in the policy. That's not true if one uses multiple policies because the choice of policy, which happens outside of the policies themselves, is security-critical. For example, if we have two policies, one used for creating links like `<a href>` and the other for setting the base URI, the code that encapsulates the function that creates links and the code that encapsulates the setting of the base URI must be audited to ensure they are using the right policy.

When I filed this, I was thinking that the solution would be to build more functionality into the browser to assist in implementing policies, and more kinds of `default`-like policies.I I had thought we could get away with using Trusted Types directly in applications, without encapsulating the usage of Trusted Types in a library. But experimenting with this shows me that's probably not going to work well as, like you suggested, having multiple policies and managing them seems to be important for using Trusted Types efficiently, and that should all be encapsulated in a library. So, I don't think there's anything actionable here.
